src/buildercore/terraform.py

```diff
@@ -564,6 +569,230 @@ def render_bigquery(context, template):
+        'role': '${aws_iam_role.worker.name}'
+    })
+
+    template.populate_data(DATA_TYPE_AWS_AMI, 'worker', block={
+        'filter': {
+            'name': 'name',
+            'values': ['amazon-eks-node-%s-v*' % context['eks']['version']],
+        },
+        'most_recent': True,
+        'owners': [aws.ACCOUNT_EKS_AMI],
+    })
+
+    # EKS currently documents this required userdata for EKS worker nodes to
+    # properly configure Kubernetes applications on the EC2 instance.
+    # We utilize a Terraform local here to simplify Base64 encoding this
+    # information into the AutoScaling Launch Configuration.
+    # More information: https://docs.aws.amazon.com/eks/latest/userguide/launch-workers.html
+    template.populate_local('worker_userdata', """
+#!/bin/bash
+set -o xtrace
+/etc/eks/bootstrap.sh --apiserver-endpoint '${aws_eks_cluster.main.endpoint}' --b64-cluster-ca '${aws_eks_cluster.main.certificate_authority.0.data}' '${aws_eks_cluster.main.name}'""")
+
+    template.populate_resource('aws_launch_configuration', 'worker', block={
+        'associate_public_ip_address': True,
+        'iam_instance_profile': '${aws_iam_instance_profile.worker.name}',
+        'image_id': '${data.aws_ami.worker.id}',
+        'instance_type': context['eks']['worker']['type'],
+        'name_prefix': '%s--worker' % context['stackname'],
+        'security_groups': ['${aws_security_group.worker.id}'],
+        'user_data_base64': '${base64encode(local.worker_userdata)}',
+        'lifecycle': {
+            'create_before_destroy': True,
+        },
+    })
+
+    autoscaling_group_tags = [{
+        'key': k,
+        'value': v,
+        'propagate_at_launch': True,
+    } for (k, v) in aws.generic_tags(context).items()]
+    autoscaling_group_tags.append({
+        'key': 'kubernetes.io/cluster/%s' % context['stackname'],
+        'value': 'owned',
+        'propagate_at_launch': True,
+    })
+    template.populate_resource('aws_autoscaling_group', 'worker', block={
+        'name': '%s--worker' % context['stackname'],
+        'launch_configuration': '${aws_launch_configuration.worker.id}',
+        'min_size': context['eks']['worker'].get('min-size', 1),
+        'max_size': context['eks']['worker'].get('max-size', 3),
+        'desired_capacity': context['eks']['worker'].get('desired-capacity', 1),
+        'vpc_zone_identifier': [context['eks']['subnet-id'], context['eks']['redundant-subnet-id']],
+        'tags': autoscaling_group_tags,
+    })
+
+    template.populate_local('config_map_aws_auth', """
+- rolearn: ${aws_iam_role.worker.arn}
+  username: system:node:{{EC2PrivateDNSName}}
+  groups:
+    - system:bootstrappers
+    - system:nodes""")
+
+    template.populate_resource('kubernetes_config_map', 'aws_auth', block={
+        'metadata': [{
+            'name': 'aws-auth',
+            'namespace': 'kube-system',
+        }],
+        'data': {
+            'mapRoles': '${local.config_map_aws_auth}',
```

Comment:
I can understand sticking a bash snippet into a var and then base64 encoding it like you have above, but what is special about this snippet that it lives in a local variable instead of a native whatever-terraform-is-using structure? is it passed directly to AWS or similar?

Reply:
It is YAML so the Terraform template looks like this:
```
    "kubernetes_config_map": {
      "aws_auth": {
        "data": {
          "mapRoles": "${local.config_map_aws_auth}"
        },
        "metadata": [
          {
            "namespace": "kube-system",
            "name": "aws-auth"
          }
        ]
      }
    }
  },
  "locals": {
    "config_map_aws_auth": "\n- rolearn: ${aws_iam_role.worker.arn}\n  username: system:node:{{EC2PrivateDNSName}}\n  groups:\n    - system:bootstrappers\n    - system:nod
es\n- rolearn: ${aws_iam_role.user.arn}\n  groups:\n    - system:masters\n"
  }
```
rather than
```
    "kubernetes_config_map": {
      "aws_auth": {
        "data": {
          "mapRoles": "\n- rolearn: ${aws_iam_role.worker.arn}\n  username: system:node:{{EC2PrivateDNSName}}\n  groups:\n    - system:bootstrappers\n    - system:nod
es\n- rolearn: ${aws_iam_role.user.arn}\n  groups:\n    - system:masters\n"
        },
        "metadata": [
          {
            "namespace": "kube-system",
            "name": "aws-auth"
          }
        ]
      }
    }
  },
```
Readability is just marginally better (when debugging the `generated.tf.json` file), but it also means unit tests are easier to write for the `kubernetes_config_map` as they can test the whole `data` and `metadata` structure without the verbose YAML embedded in it.